clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedCommand.java

```diff
@@ -31,190 +40,308 @@
+            }
+        } else if (message.getReplyTo() != null) {
+            // correlation id is required if a command response is expected
+            validationErrorJoiner.add("neither message-id nor correlation-id is set");
+        }
+
+        String replyToId = null;
+        if (message.getReplyTo() != null) {
+            try {
+                final ResourceIdentifier replyTo = ResourceIdentifier.fromString(message.getReplyTo());
+                if (!CommandConstants.isNorthboundCommandResponseEndpoint(replyTo.getEndpoint())) {
+                    validationErrorJoiner.add("reply-to not a command address: " + message.getReplyTo());
+                } else if (tenantId != null && !tenantId.equals(replyTo.getTenantId())) {
+                    validationErrorJoiner.add("reply-to not targeted at tenant " + tenantId + ": " + message.getReplyTo());
+                } else {
+                    replyToId = replyTo.getPathWithoutBase();
+                    if (replyToId.isEmpty()) {
+                        validationErrorJoiner.add("reply-to part after tenant not set: " + message.getReplyTo());
+                    }
+                }
+            } catch (final IllegalArgumentException e) {
+                validationErrorJoiner.add("reply-to cannot be parsed: " + message.getReplyTo());
+            }
+        }
+        return new ProtonBasedCommand(
+                validationErrorJoiner.length() > 0 ? Optional.of(validationErrorJoiner.toString()) : Optional.empty(),
+                message,
+                tenantId,
+                deviceId,
+                correlationId,
+                replyToId);
     }
 
     /**
-     * {@inheritDoc}
+     * Creates a command from an AMQP 1.0 message.
+     * <p>
+     * The command is created as described in {@link #from(Message)}.
+     * Additionally, gateway information is extracted from the <em>via</em> property of the message.
+     *
+     * @param message The message containing the command.
+     * @return The command.
+     * @throws NullPointerException if message is {@code null}.
      */
-    @Override
-    public boolean isOneWay() {
-        return command.isOneWay();
+    public static ProtonBasedCommand fromRoutedCommandMessage(final Message message) {
+        final ProtonBasedCommand command = from(message);
+        final String gatewayId = MessageHelper.getApplicationProperty(message.getApplicationProperties(),
+                MessageHelper.APP_PROPERTY_CMD_VIA, String.class);
+        if (!Strings.isNullOrEmpty(gatewayId)) {
+            command.setGatewayId(gatewayId);
+        }
+        return command;
     }
 
     /**
-     * {@inheritDoc}
+     * Sets the identifier of the gateway this command is to be sent to.
+     *
+     * @param gatewayId The gateway identifier.
      */
+    public void setGatewayId(final String gatewayId) {
```

Comment:
why do we need a setter for the gateway ID? Isn't it determined from the message?

Reply:
See above - same applies to the Kafka record.